pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -375,8 +375,17 @@ func (ssc *defaultStatefulSetControl) processReplica(
 	replicas []*v1.Pod,
 	i int) (bool, error) {
 	logger := klog.FromContext(ctx)
-	// delete and recreate failed pods
-	if isFailed(replicas[i]) {
+	// Delete and recreate pods which finished running.
+	//
+	// Note that pods with phase Succeeded will also trigger this event. This is
+	// because final pod phase of evicted or otherwise forcibly stopped pods
+	// (e.g. terminated on node reboot) is determined by the exit code of the
+	// container, not by the reason for pod termination. We should restart the pod
+	// regardless of the exit code.
+	//
+	// Also note that it is impossible for StatefulSet pods to ever truly complete,
+	// as validation allows only restartPolicy=Always.
+	if isFinished(replicas[i]) {
 		ssc.recorder.Eventf(set, v1.EventTypeWarning, "RecreatingFailedPod",
```

Comment:
I think it is better to issue a dedicated event `RecreatingSucceededPod`. This is what we do for daemon set in the analogous case: https://github.com/kubernetes/kubernetes/pull/117073/files

Reply:
Potentially, but I'm concerned about backwards compatibility as this would mean renaming an existing event type. I would suggest decoupling this to a dedicated discussion / PR. @alculquicondor wdyt?

Reply:
Based on discussion on sig-apps slack:
- we need to be careful with renaming events of type warning (so, `RecreateFailedPod` stays as it is),
- just recreating a pod should be emitted as a normal event, not warning.

@alculquicondor DaemonSet events are already different - I might be missing something, but I haven't found any 'recreate' events, only `FailedDaemonPod` and `SucceededDaemonPod` informing of deleting the pod. So I'm not sure if we need to name the new event `RecreateSucceededPod`, as there's already little consistency. I think we can use `RecreateTerminatedPod` here, and it'll be much clearer to the user.